File: android/app/src/test/java/com/booksync/player/PlaybackOffsetsTest.kt
package com.booksync.player

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * The two playback offsets (issue #42).
 *
 * The bug these pin: the same button used to move a different distance on every
 * surface — 15s on the phone player, 10s in Android Auto, 15s back / 30s forward
 * on the web, and a 2s jump on the reader's text→audio handoff against a stated
 * server default of 10s. The values are asserted literally here so a change on
 * one platform can't quietly desync from the others; the matching web copy lives
 * in `AudioPlayerContext.jsx` and the server's in `config.default_rewind_seconds`.
 */
class PlaybackOffsetsTest {

    @Test
    fun `skip is 30 seconds and resume rewind is 5`() {
        assertEquals(30_000L, PlaybackOffsets.SKIP_MS)
        assertEquals(5_000L, PlaybackOffsets.RESUME_REWIND_MS)
    }

    @Test
    fun `resume backs up by the rewind offset`() {
        assertEquals(85_000L, PlaybackOffsets.resumePosition(90_000L))
    }

    @Test
    fun `resume clamps at the start of the file`() {
        assertEquals(0L, PlaybackOffsets.resumePosition(2_000L))
        assertEquals(0L, PlaybackOffsets.resumePosition(0L))
    }

    @Test
    fun `skip back moves one skip offset`() {
        assertEquals(70_000L, PlaybackOffsets.skipBackPosition(100_000L))
    }

    @Test
    fun `skip back clamps at the start of the file`() {
        assertEquals(0L, PlaybackOffsets.skipBackPosition(10_000L))
    }

    @Test
    fun `skip forward moves one skip offset`() {
        assertEquals(130_000L, PlaybackOffsets.skipForwardPosition(100_000L, durationMs = 600_000L))
    }

    @Test
    fun `skip forward clamps at the end of the file`() {
        assertEquals(600_000L, PlaybackOffsets.skipForwardPosition(590_000L, durationMs = 600_000L))
    }

    @Test
    fun `skip forward ignores an unknown duration rather than seeking backwards`() {
        // ExoPlayer reports C.TIME_UNSET (a large negative) before the media is
        // prepared. Coercing against that would drag the position backwards.
        assertEquals(130_000L, PlaybackOffsets.skipForwardPosition(100_000L, durationMs = -9_223_372_036_854_775_807L))
    }
}

File: web/src/lib/playbackOffsets.test.js
import { describe, it, expect } from 'vitest'
import { SKIP_SECONDS, RESUME_REWIND_SECONDS, handoffPositionMs } from './playbackOffsets'

// The web half of the cross-platform offsets contract (issue #42). Android's
// PlaybackOffsetsTest and the server's
// test_epub_to_audio_exact_match_applies_default_rewind assert the same two
// numbers literally, so a one-sided edit fails CI instead of shipping a silent
// divergence. See docs/position-sync-contract.md § Playback offsets.
describe('playback offsets', () => {
    it('skips 30s and rewinds 5s, the same numbers Android and the server use', () => {
        expect(SKIP_SECONDS).toBe(30)
        expect(RESUME_REWIND_SECONDS).toBe(5)
    })
})

describe('handoffPositionMs (issue #212)', () => {
    it('lands the resume rewind before the anchor', () => {
        expect(handoffPositionMs(42000)).toBe(37000)
    })

    it('clamps at 0 rather than seeking negative near the start of the book', () => {
        expect(handoffPositionMs(2000)).toBe(0)
        expect(handoffPositionMs(0)).toBe(0)
    })

    it('treats a missing anchor as the start of the book', () => {
        expect(handoffPositionMs(null)).toBe(0)
        expect(handoffPositionMs(undefined)).toBe(0)
    })
})
